### PR TITLE
ARROW-3363: [C++/Python] Add helper functions to detect scalar Python types

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

**Comment by xhochy** [comment]
Required for https://github.com/apache/arrow/pull/2623

**Comment by kszucs** [comment]
@xhochy clang-format fails

File: python/pyarrow/types.pxi [hunk 1]
```diff
@@ -1474,3 +1474,15 @@ def from_numpy_dtype(object dtype):
         check_status(NumPyDtypeToArrow(dtype, &c_type))
 
     return pyarrow_wrap_data_type(c_type)
+
+
+def is_boolean_object(object obj):
```
Comment (kszucs) [inline python/pyarrow/types.pxi]: The naming and/or placement in types module feels a bit odd.

```python
pa.types.is_integer_object(pa.types.int16())  # False
```

Perhaps `is_<type>_value`? Or how about creating a `pyarrow.inference` module with `is_bool`, `is_int`, `is_float` functions?

Comment (wesm) [reply python/pyarrow/types.pxi]: I like the _value name ok

**Comment by xhochy** [comment]
> @xhochy clang-format fails
 Once https://github.com/apache/arrow/pull/2572 is merged, I'll add a `docker-compose run clang-format`. It's really annoying that this continuously breaks for me when brew updates clang-format.

**Review by kszucs — APPROVED** [review]
LGTM

**Comment by kszucs** [comment]
Sounds good, to run `docker-compose run clang-format` arrow must be attached as a volume though.

**Comment by kszucs** [comment]
CI failure is related to wheel incompatibilities.

**Pull request closed without merging**